httpd/streaming.c

```diff
@@ -0,0 +1,378 @@
+    rbuf_free(conn->tx);
+
+    freez(conn->url);
+    freez(conn->user_agent);
+
+    pthread_mutex_destroy(&conn->rx_buf_lock);
+    pthread_mutex_destroy(&conn->tx_buf_lock);
+    pthread_cond_destroy(&conn->rx_buf_cond);
+}
+
+// streaming upgrade related functions
+int is_streaming_handshake(h2o_req_t *req)
+{
+    /* method */
+    if (!h2o_memis(req->input.method.base, req->input.method.len, H2O_STRLIT("GET")))
+        return 1;
+
+    if (!h2o_memis(req->path_normalized.base, req->path_normalized.len, H2O_STRLIT(NETDATA_STREAM_URL))) {
+        return 1;
+    }
+
+    /* upgrade header */
+    if (req->upgrade.base == NULL || !h2o_lcstris(req->upgrade.base, req->upgrade.len, H2O_STRLIT(NETDATA_STREAM_PROTO_NAME)))
+        return 1;
+
+    // TODO consider adding some key in form of random number
+    // to prevent caching on route especially if TLS is not used
+    // e.g. client sends random number
+    // server replies with it xored
+
+    return 0;
+}
+
+static void stream_on_close(h2o_stream_conn_t *conn);
+void stream_process(h2o_stream_conn_t *conn, int initial);
+
+void stream_on_complete(void *user_data, h2o_socket_t *sock, size_t reqsize)
+{
+    h2o_stream_conn_t *conn = user_data;
+
+    /* close the connection on error */
+    if (sock == NULL) {
+        stream_on_close(conn);
+        return;
+    }
+
+    conn->sock = sock;
+    sock->data = conn;
+
+    conn_list_insert(&conn_list, conn);
+
+    h2o_buffer_consume(&sock->input, reqsize);
+    stream_process(conn, 1);
+}
+
+// handling of active streams
+static void stream_on_close(h2o_stream_conn_t *conn)
+{
+    if (conn->sock != NULL)
+        h2o_socket_close(conn->sock);
+
+    conn_list_remove_conn(&conn_list, conn);
+
+    pthread_mutex_lock(&conn->rx_buf_lock);
+    conn->shutdown = 1;
+    pthread_cond_broadcast(&conn->rx_buf_cond);
+    pthread_mutex_unlock(&conn->rx_buf_lock);
+
+    h2o_stream_conn_t_destroy(conn);
+    freez(conn);
+}
+
+static void on_write_complete(h2o_socket_t *sock, const char *err)
+{
+    h2o_stream_conn_t *conn = sock->data;
+
+    if (err != NULL) {
+        stream_on_close(conn);
+        error_report("Streaming connection error \"%s\"", err);
+        return;
+    }
+
+    pthread_mutex_lock(&conn->tx_buf_lock);
+
+    rbuf_bump_tail(conn->tx, conn->tx_buf.len);
+
+    conn->tx_buf.base = NULL;
+    conn->tx_buf.len = 0;
+
+    pthread_mutex_unlock(&conn->tx_buf_lock);
+
+    stream_process(conn, 0);
+}
+
+static void stream_on_recv(h2o_socket_t *sock, const char *err)
+{
+    h2o_stream_conn_t *conn = sock->data;
+
+    if (err != NULL) {
+        stream_on_close(conn);
+        error_report("Streaming connection error \"%s\"", err);
+        return;
+    }
+    stream_process(conn, 0);
+}
+
+#define PARSE_DONE 1
+#define PARSE_ERROR -1
+#define GIMME_MORE_OF_DEM_SWEET_BYTEZ 0
+
+#define STREAM_METHOD "STREAM "
+#define HTTP_1_1 " HTTP/1.1"
+#define HTTP_HDR_END "\r\n\r\n"
+#define USER_AGENT "User-Agent: "
```

Comment:
Should we add these constants inside a header? :thinking: 
They could be used with other files  as well.